### DISCUSSION TITLE
printer() in custom serializer triggers out of memory error with any value

### DISCUSSION THREAD

**Discussion by gterras** [vitest-dev/vitest#5751] [Q&A] [answered]
Hi following https://vitest.dev/guide/snapshot.html#custom-serializer I have: 

```ts
// custom-serializer.ts
import type { SnapshotSerializer } from 'vitest'

export default {
	serialize(_val, config, indentation, depth, refs, printer) {
		return printer('simple string', config, indentation, depth, refs) // notice simple value 
	},
	test() {
		return true
	},
} satisfies SnapshotSerializer
```
```ts
// vitest.config.mjs 
export default defineConfig({
  test: {
    plugins: [swc.vite({ module: { type: 'es6' } }), tsconfigPaths()],
    snapshotSerializers: ['./custom-serializer.ts'],
    ...
    }
}
```

Now any running test will trigger an out of memory error: 
```ts
expect(res)toMatchSnapshot()
```

```bash
<--- Last few GCs --->

[1828836:0x7fbe24000f00]     4079 ms: Scavenge 149.7 (171.1) -> 139.3 (172.1) MB, 1.71 / 0.00 ms  (average mu = 0.995, current mu = 0.997) allocation failure; 
[1828836:0x7fbe24000f00]     4436 ms: Scavenge 154.9 (175.2) -> 145.0 (178.7) MB, 1.95 / 0.01 ms  (average mu = 0.995, current mu = 0.997) allocation failure; 
[1828836:0x7fbe24000f00]     4514 ms: Scavenge 158.8 (179.5) -> 147.8 (182.7) MB, 1.43 / 0.01 ms  (average mu = 0.995, current mu = 0.997) task; 


<--- JS stacktrace --->

FATAL ERROR: RegExpCompiler Allocation failed - process out of memory
----- Native stack trace -----

 1: 0xb84bd6 node::OOMErrorHandler(char const*, v8::OOMDetails const&) [node (vitest)]
 2: 0xefec30 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, v8::OOMDetails const&) [node (vitest)]
 3: 0xefef17 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, v8::OOMDetails const&) [node (vitest)]
 4: 0x14fc958  [node (vitest)]
 5: 0x14eaaf0 v8::internal::RegExpAlternative::ToNode(v8::internal::RegExpCompiler*, v8::internal::RegExpNode*) [node (vitest)]
 6: 0x14e8ef9 v8::internal::RegExpCapture::ToNode(v8::internal::RegExpTree*, int, v8::internal::RegExpCompiler*, v8::internal::RegExpNode*) [node (vitest)]
 7: 0x14fe6bd v8::internal::RegExpCompiler::PreprocessRegExp(v8::internal::RegExpCompileData*, v8::base::Flags<v8::internal::RegExpFlag, int>, bool) [node (vitest)]
 8: 0x15226cf v8::internal::RegExpImpl::Compile(v8::internal::Isolate*, v8::internal::Zone*, v8::internal::RegExpCompileData*, v8::base::Flags<v8::internal::RegExpFlag, int>, v8::internal::Handle<v8::internal::String>, v8::internal::Handle<v8::internal::String>, bool, unsigned int&) [node (vitest)]
 9: 0x1522fd6 v8::internal::RegExpImpl::CompileIrregexp(v8::internal::Isolate*, v8::internal::Handle<v8::internal::JSRegExp>, v8::internal::Handle<v8::internal::String>, bool) [node (vitest)]
10: 0x1523b97 v8::internal::RegExpImpl::IrregexpPrepare(v8::internal::Isolate*, v8::internal::Handle<v8::internal::JSRegExp>, v8::internal::Handle<v8::internal::String>) [node (vitest)]
11: 0x1523cd0 v8::internal::RegExpImpl::IrregexpExec(v8::internal::Isolate*, v8::internal::Handle<v8::internal::JSRegExp>, v8::internal::Handle<v8::internal::String>, int, v8::internal::Handle<v8::internal::RegExpMatchInfo>, v8::internal::RegExp::ExecQuirks) [node (vitest)]
12: 0x154c835 v8::internal::Runtime_StringReplaceNonGlobalRegExpWithFunction(int, unsigned long*, v8::internal::Isolate*) [node (vitest)]
13: 0x1971ef6  [node (vitest)]
Aborted (core dumped)
```

When directly returning a value from `serialize()` without using `printer()` this works as expected. Also note that using the snapshot feature without custom parser works flawlessly even on really complex objects. 

The same also happens when directly adding the serializer in the test:
```ts
expect.addSnapshotSerializer({
	serialize(val, config, indentation, depth, refs, printer) {
	    return printer('simple string', config, indentation, depth, refs) // notice simple value 	
	 },
	test(val) {
		return !!val
	},
})

expect(res).toMatchSnapshot() // same error 
```

Am I missing something? 

Thanks for your answers.

**Comment by hi-ogawa** [comment]
If I remember, `printer` will recurse all custom serializer, so I would I assume it's looping (I thought that would be stack overflow, but maybe it concatenates string faster and go in to heap error first).

I assume your example is just for the sake of crash and wouldn't be useful anyway. Can you show what you were trying to do in the first place (which probably lead you to discover something odd like this)?

---

I think I got this loop myself in the past when making custom snapshot, so maybe it's worth adding some notes in the doc, but I think it's somewhat "obvious" that'll happen, so I'm not sure what's the better way to explain it.

**Reply by gterras** [reply]
> I assume your example is just for the sake of crash and wouldn't be useful anyway. Can you show what you were trying to do in the first place (which probably lead you to discover something odd like this)?

I don't really understand as I'm just trying to replicate the example from the docs. The problem occurs even when using `val` (so by strictly copy-pasting the example).

**Reply by hi-ogawa** [reply]
The example from the doc should be working fine
https://stackblitz.com/edit/vitest-dev-vitest-ef4kgt?file=test%2Frepro.test.ts

so you probably misunderstood something and getting into the loop. What I meant was that doing like below would call into your serializer forever so it should crash:

```ts
  expect.addSnapshotSerializer({
    serialize(val, config, indentation, depth, refs, printer) {
      // `serialize(val, ...)` will get called forever since `test(val) = true` forever
      return printer(val, config, indentation, depth, refs)
    },
    test(val) {
      return val && Object.prototype.hasOwnProperty.call(val, 'foo');
    },
  });
```

**Reply by gterras** [reply]
Thanks for your answer, with this information it started to be more clear that `printer()` isn't was I think it was: I though it was a simple helper when it's actually an advanced feature that needs a lot of care (notably regarding recursion) to be used properly (and is probably isn't needed anyway when you don't have to write/interact with plugins). 

Initially my goal was to tweak the output of snap files because I didn't like what Vitest did with the output of string (adding a double quote on the same line as first character) since it was messing with my `.md` table headers : 

vanilla: 

```ts
// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test name`] = `
"| Date       | Capital |
| :--------- | :------ | 
| 01/01/2020 | 1000000 | 
| 01/01/2021 | 1000000 | 
"
`;
```

So I've dig into custom serializers to get this output: 

```ts
// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test name`] = `
| Date       | Capital | 
| :--------- | :------ | 
| 01/01/2020 | 1000000 | 
| 01/01/2021 | 1000000 | 
`;
```

Which now I can do with: 

```ts
export default {
	serialize(val, ) {
		return val
	},
	test(val) {
		return val && typeof val === 'string'
	},
} satisfies SnapshotSerializer
```


I naively thought that overriding the serializer implied to mandatory call printer, mostly because that's how the docs introduce it, but also because I thought the serializer was responsible for the whole `.snap` file when in reality it is scoped to the current test output - which is an awesome feature that I don't think is told in the docs, ie running two `expect(res).toMatchSnapshot()` in the same test will gracefully result in: 

```js
exports[`testname 1`] = `...`
exports[`testname 2`] = `...`
```

Now what I understand is that there is no way to override the serializer for a single `expect` within a test? IE: 

```js
it('should pass',  ({expect}) => {
expect(table).toMatchSnapshot() // custom serializer
expect(object).toMatchSnapshot() // default serializer 
})
```

Something like `expect(table, custom_serializer).toMatchSnapshot()`, or a way to call default implementation from the custom serializer in some cases?

In the end all I need is a simple way to override the outputted value on demand for any given expect.

**Reply by hi-ogawa** [reply, marked as answer]
> Now what I understand is that there is no way to override the serializer for a single `expect` within a test? IE:
> 
> ```js
> it('should pass',  ({expect}) => {
> expect(table).toMatchSnapshot() // custom serializer
> expect(object).toMatchSnapshot() // default serializer 
> })
> ```
> 
> Something like `expect(table, custom_serializer).toMatchSnapshot()`, or a way to call default implementation from the custom serializer in some cases?
>
> In the end all I need is a simple way to override the outputted value on demand for any given expect.

For this need, introducing a simple wrapper might help, something like this:
https://stackblitz.com/edit/vitest-dev-vitest-w9sfer?file=test%2Frepro.test.ts

```ts
import { test, expect } from 'vitest';

class MySnapshotWrapper {
  constructor(public inner: string) {}
}

expect.addSnapshotSerializer({
  test(val) {
    return val instanceof MySnapshotWrapper;
  },
  serialize(val, config, indentation, depth, refs, printer) {
    // you can also tweak anything here e.g. reusing `printer` first then modify output
    return (val as MySnapshotWrapper).inner;
  },
});

test('repro', () => {
  expect(new MySnapshotWrapper('my\nthing')).toMatchInlineSnapshot(`
    my
    thing
  `);
  expect('other\nthing').toMatchInlineSnapshot(`
  "other
  thing"
`);
});
```

I've used this idea on my own in the past as I wanted to mostly reuse the default `printer` output, but just needed to tweak the output to preserve whitespace the way it makes more sense https://github.com/hi-ogawa/js-utils/blob/e38af9ce06108056b85ec553b0f090610950a599/packages/tiny-react/src/ssr/index.test.tsx#L278-L293

**Reply by gterras** [reply]
I see thanks for your answer. I think I'll stick with `toMatchFileSnapshot` even though it means losing the nice features of `toMatchSnapshot` like test names/hints headers or single file output since I would like to stick to Vitest basic API. 

Some kind of simple `transform` function would be a nice addition to `toMatchSnapshot` to offer control on the output without the hassle of using custom serializers.
